maps-compose/src/main/java/com/google/maps/android/compose/GoogleMap.kt

```diff
@@ -123,107 +137,224 @@ public fun GoogleMap(
+                            if(targetState < currentLifecycleState) moveBackward()
+                            else moveForward()
+                        }
+                    }
+
+                    private fun moveBackward() {
+                        val event = Lifecycle.Event.downFrom(currentLifecycleState)
+                            ?: error("no event down from $currentLifecycleState")
+                        invokeEvent(event)
+                    }
+
+                    private fun moveForward() {
+                        val event = Lifecycle.Event.upFrom(currentLifecycleState)
+                            ?: error("no event up from $currentLifecycleState")
+                        invokeEvent(event)
+                    }
+
+                    private fun invokeEvent(event: Lifecycle.Event) {
+                        log("Invoking Lifecycle event $event")
+                        when(event) {
+                            Lifecycle.Event.ON_CREATE -> mapView.onCreate(Bundle())
+                            Lifecycle.Event.ON_START -> mapView.onStart()
+                            Lifecycle.Event.ON_RESUME -> mapView.onResume()
+                            Lifecycle.Event.ON_PAUSE -> mapView.onPause()
+                            Lifecycle.Event.ON_STOP -> mapView.onStop()
+                            Lifecycle.Event.ON_DESTROY -> mapView.onDestroy()
+                            else -> error("Unsupported lifecycle event: $event")
+                        }
+                        currentLifecycleState = event.targetState
+                    }
                 }
+
+                var lifecycleOwner by Delegates.notNull<LifecycleOwner>()
+
+                fun unregisterLifecycleObserver() {
+                    log("Unregistering lifecycle observer")
+                    lifecycleOwner.lifecycle.removeObserver(lifecycleObserver)
+                }
+
+                val attachStateListener = object : View.OnAttachStateChangeListener {
+                    override fun onViewAttachedToWindow(v: View) {
+                        log("View attached!")
+                        lifecycleOwner = v.findViewTreeLifecycleOwner()!!
+                        lifecycleOwner.lifecycle.addObserver(lifecycleObserver)
+                    }
+
+                    override fun onViewDetachedFromWindow(v: View) {
+                        log("View detached!")
+                        unregisterLifecycleObserver()
+                        lifecycleObserver.moveToLifecycleState(Lifecycle.State.CREATED)
+                    }
+                }
+
+                mapView.addOnAttachStateChangeListener(attachStateListener)
+
+                mapView.tagData().onRelease = {
+                    unregisterLifecycleObserver()
+                    mapView.removeOnAttachStateChangeListener(attachStateListener)
+                    lifecycleObserver.moveToLifecycleState(Lifecycle.State.DESTROYED)
+                }
```

Comment:
Couple small things here: Personally I'd prefer to store the custom LifecycleEventObserver in the tag, not a lambda. I think it's clearer and cleaner. I don't think `removeOnAttachStateChangeListener()` needs to be called here, it will go away with the View. Similar for `unregisterLifecycleObserver()`, it would be called from onViewDetachedFromWindow(), and if it did not it would mean that the Activity is being destroyed or similar, meaning everything is being GCed anyway.

The only thing that I think we cannot get rid of here is moving to DESTROYED state. I would create a custom method for this, similar to "moveToBaseState()", because, once again, I think we should deal with that theoretical case of never getting out of INITIALIZED state (by checking the current state), meaning we never even called MapView.onCreate(). I see this special case as the only hard reason why we need access to the LifecycleEventObserver at all from onRelease; it tracks the current MapView lifecycle state.

Also, I'm not 100% sure if `onViewDetachedFromWindow()` will always run before onRelease, perhaps there could be a race flipping the order. Because of this I would change the error() calls in moveBackward() and moveForward() to logged warnings of some sort, then nothing bad will happen if we're already DESTROYED when onViewDetachedFromWindow wants to move to CREATED. Or you could create an issuetracker case requesting clarification of the `AndroidView()` documentation.

Reply:
I guess the current state check in `moveToBaseState()` can take care of the DESTROYED vs. CREATED confusion & race, right? No need to remove error() calls.

Reply:
This should be taken care of by checking if `currentLifecycleState > Lifecycle.State.CREATED` before moving to `Lifecycle.State.CREATED` because `Lifecycle.State.DESTROYED` is below/less than `Lifecycle.State.CREATED`.

Reply:
I created a moveToDestroyedState method.